### ISSUE TITLE
Add syntax highlighting to website build

### ISSUE THREAD

**Issue by morganwillcock** [adventuregamestudio/ags-manual#200]
It should be possible to just statically include enough of the Prism syntax highlighter to highlight the AGS scripting language and diaglog scripts, either from the [WIP branch](https://github.com/adventuregamestudio/prism/tree/ags) or (hopefully at some point) using the upstream version of Prism.

As a quick test, you can run the Prism site locally, just download the required parts from [http://127.0.0.1:8080/download.html#themes=prism&languages=clike+ags+agsdialog](http://127.0.0.1:8080/download.html#themes=prism&languages=clike+ags+agsdialog), and then use it like this:
```html
<!DOCTYPE html>
<html>
<head>
  <link href="prism.css" rel="stylesheet" />
</head>
<body>
  <script src="prism.js"></script>
  <pre><code class="language-agsdialog">@1
ego: "Hello. How are you?"
narrator: The man looks you in the eye.
  if (player.HasInventory(iKey)) {
    player.Say("Actually, I'd better go.");
    return RUN_DIALOG_STOP_DIALOG;
  }
otherman: "Here's a key for you."
return
</code></pre>
</body>
</html>
```

This will likely require:
* switching the code blocks in the Markdown source to use backticks and a language identifier
* rewriting of the language identifer during export to ensure that it is prefixed with "language-" (this is apparently not straight forward and Pandoc does not do this as the default behaviour when its own highlighing mechanism is disabled)

**Comment by morganwillcock** [comment]
As an alternative that doesn't use Prism. Pandoc supports using KDE syntax highlighting definitions to implement its highlighting, and that allows dynamically switching in and out of different languages.

It looks possible to build everything statically; I tested it like this:
```sh
pandoc -t html5 --standalone --syntax-definition=agsdialog.xml --syntax-definition=ags.xml -o test.html test.md
```
...and it does allow the switching between those two definitions.

Disadvantages of this way:
* The theme of the highlighting is determined at build time, so potentially it is less flexible when the user wants to dynamically switch the theme
* It duplicates the existing work on the Prism highlighter

Advantages of this way:
* The syntax matching is designed for an editor so it should be possible to implement matching in a more correct way with greater awareness of context
* Highlighting will work without running any additional scripts on the user side, so it would be more reliable and safer
* Some KDE based editors would potentially get some language support for free

**Comment by AlanDrake** [comment]
Couldn't we use cpp code blocks directly?
Might not be enough for dialogs, but for the rest of the code should be good enough and work right off the bat on github and pandoc, we just have to replace the current indented blocks, with actual code blocks that start with \```cpp

**Comment by morganwillcock** [comment]
I actually already wrote the syntax highlighting definitions that would work natively and render the highlighting directly into the pages: [syntax.zip](https://github.com/adventuregamestudio/ags-manual/files/9788955/syntax.zip)

The problem is how this interacts with the light and dark theme switching: effectively all the CSS needs re-doing to implement something that matches an existing highlight style, or it needs a (light and dark) custom highlight style, or both of those.

It would probably work in the CHM content too (I think it is effectively rendered in IE7), but the pages are built separately so they don't necessarily have to be changed.

**Comment by ericoporto** [comment]
> I actually already wrote the syntax highlighting definitions that would work natively and render the highlighting directly into the pages

Hey, it may be useful to store this in the AGS org since it can be useful for other things. Not sure if it's own repo or some AGS script language grammar or utility repository.

**Comment by morganwillcock** [comment]
I was wondering what to do with them. I used Kate for testing: as pure syntax highlighters they seem to work pretty well. It wasn't quite as flexible as I was hoping for because the lookahead seems to effectively be limited to one syntax match, but I think it should be at least as good as the Prism highlighting.

**Comment by morganwillcock** [comment]
I've pushed what I already did to a separate branch:
https://github.com/adventuregamestudio/ags-manual/tree/syntax-highlighting

Looking at the build output of that is probably worthwhile to see the effect of actually adding the highlighting to all instances of code (this is done using a Lua filter). I didn't include the custom theme that I started, instead it is set to use the "pygments" theme.

**Comment by ericoporto** [comment]
> I didn't include the custom theme that I started, instead it is set to use the "pygments" theme.

I can't seem to figure this out, where is this "pygments" theme? Can we use it as a base to create a newer theme? If it is in css perhaps it could use some variables so it has dark and light variants.

**Comment by AlanDrake** [comment]
It doesn't seem to be working perfectly.
Check StringFormats.html for example, some tokens are correctl colored, others not.

**Comment by morganwillcock** [comment]
> I can't seem to figure this out, where is this "pygments" theme? Can we use it as a base to create a newer theme? If it is in css perhaps it could use some variables so it has dark and light variants.

It is built into Pandoc, there are a few different options. It isn't CSS: print one as an example:
```sh
pandoc --print-highlight-style pygments
```

> Or is this only for the jsless chm web build?

It doesn't need JavaScript, so it would work in both.

> It doesn't seem to be working perfectly.
> Check StringFormats.html for example, some tokens are correctly colored, others not.

I guess it is a Pandoc issue. The same block highlights fine when I test it.
![kate](https://user-images.githubusercontent.com/14958747/196000813-fde2eb52-0014-435f-9981-9a2263f54300.png)

**Comment by AlanDrake** [comment]
I tried locally with latest pandoc and without the extra lua, by using fenced blocks, and that way the sytnax works as expected. Maybe we just need to update pandoc, unless it was `set_default_code_class.lua` causing the issue.

Look, let's do it like this: get rid of that `set_default_code_class.lua` which will probably break some stuff.
I'll just go and replace all code blocks with fenced ones with the correct language (ags and agsdialog), this way will work both for github (which appears to support "ags" syntax for some reason).

**Comment by ericoporto** [comment]
GitHub has AGS script in somewhere, I believe edmundo added a long time ago.

**Comment by AlanDrake** [comment]
It was a pain, but I finished normalizing all code blocks, and admonitions too while I was at it.
I left generic monospace blocks as they were, since we could keep that as a distinction from actual code.

So:
- ` ```ags ` for script code blocks
- ` ```agsdialog ` for dialog scripts
- Four spaces for common monospace content

      I am monospaced


I better be getting one of those lifetime awards next time, for freshening up the docs and the Editor UI. :smile:

@morganwillcock: you can remove `set_default_code_class.lua` and try again with the latest sources. If it still doesn't highlitght correctly then it's pandoc who needs updating.

**Comment by morganwillcock** [comment]
I think it still doesn't work, so I guess the problem is the Pandoc version. From the release notes it might be fixed in version 2.15:

> Use skylighting-0.12, skylighting-core-0.12. This fixes highlighting issues with typescript, scala, and other syntaxes that include keyword lists from different syntaxes.

I'll try to find which version it is and how to test for it.

**Comment by morganwillcock** [comment]
2.11 to fix the syntax highlighting, and then 2.11.1.1 to fix the link parsing that is broken in 2.11.
I've updated the branch.

**Comment by morganwillcock** [comment]
This was implemented and released in version 0.8.7. It it looks bad or has highlighting errors, those are separate issue.
